File: Asmi Bhaskar/day 29.c
#include <stdio.h>
#include <stdlib.h>

void update(int* BIT, int index, int value, int n) {
    for (; index <= n; index += index & -index) {
        BIT[index] += value;
    }
}

int getSum(int* BIT, int index) {
    int sum = 0;
    for (; index > 0; index -= index & -index) {
        sum += BIT[index];
    }
    return sum;
}

int countInversions(int* arr, int n) {
    int* temp = (int*)malloc(n * sizeof(int));
    for (int i = 0; i < n; i++) {
        temp[i] = arr[i];
    }

    qsort(temp, n, sizeof(int), compare);

    for (int i = 0; i < n; i++) {
        arr[i] = (int)(lower_bound(temp, n, arr[i]) - temp) + 1;
    }

    int* BIT = (int*)calloc(n + 1, sizeof(int));
    int invCount = 0;

    for (int i = n - 1; i >= 0; i--) {
        invCount += getSum(BIT, arr[i] - 1);
        update(BIT, arr[i], 1, n);
    }

    free(temp);
    free(BIT);

    return invCount;
}

int* lower_bound(int* arr, int size, int value) {
    int* low = arr;
    int* high = arr + size;
    while (low < high) {
        int* mid = low + (high - low) / 2;
        if (*mid < value) {
            low = mid + 1;
        } else {
            high = mid;
        }
    }
    return low;
}

int compare(const void* a, const void* b) {
    return (*(int*)a - *(int*)b);
}

int main() {
    int arr[] = {8, 4, 2, 1};
    int n = sizeof(arr) / sizeof(arr[0]);

    int result = countInversions(arr, n);
    printf("Number of inversions: %d\n", result);

    return 0;
}

File: Princy_Ballabh/Day_29.java
public class Day_29 
{ 
    public static void main(String args[]) 
    { 
        int arr[]={8,4,2,1};  
        System.out.println("Number of inversions are : "+getcount(arr,arr.length)); 
    } 

    static int getcount(int arr[], int n) 
    { 
        int count=0,max=0,i;  
        for(i=0;i<n;i++) 
            if(max<arr[i]) 
                max=arr[i];
        int BIT[]=new int[max+1]; 
        for(i=1;i<=max;i++) 
            BIT[i]=0; 
        for(i=n-1;i>=0;i--) 
        { 
            count+=getsum(BIT,arr[i]-1); 
            update(BIT,max,arr[i],1); 
        } 
        return count; 
    } 

    static int getsum(int BITree[], int index) 
    { 
        int sum=0; 
        while(index>0) 
        { 
            sum+=BITree[index]; 
            index-=index&(-index); 
        } 
        return sum; 
    } 

    static void update(int[] BITree, int n, int index, int val) 
    { 
        while(index<=n) 
        { 
            BITree[index]+=val; 
            index+=index&(-index); 
        } 
    } 
}
